puzzle-utils/words/ladder.py

```diff
@@ -0,0 +1,148 @@
+from .search import search
+
+def ladder(word):
+    """
+    Given a word, find all "adjacent" words in the dictionary
+    i.e. all words with edit distance of 1 from the given word
+
+    Params:
+    word (required) - The base word
+
+    Alternatives:
+    find_ladders        - Find all ladders between two words
+    find_ladders_pretty - Print all unique ladders between two words nicely
+    """
+    found_words = set()
+    for i in range(len(word)):
+        pattern = list(word)
+        pattern[i] = '.'
+        search_results = search("^" + "".join(pattern) + "$")
+        for result in search_results:
+            if result != word and result not in found_words:
+                found_words.add(result)
+    return found_words
+
+def edit_distance(left_word, right_word):
+    """
+    This assumes the lengths of the words are the same.
+    The result is the edit distance of the two words (number of
+    substitutions required to tranform left_word into right_word).
+    """
+    if len(left_word) != len(right_word):
+        raise ValueError("Word ladder words must be same length")
+
+    distance = 0;
+    for i in range(len(left_word)):
+        if left_word[i] != right_word[i]:
+            distance += 1
+    return distance
+
+def follow_ladder_path(word_path, right_word, cached_searches, depth, exact_depth, max_depth, use_heuristic):
```

Comment:
I get the impression that users would always provide an `exact_depth` and set `use_heuristic` to `true`. What are the use cases of the other flavors of calling this function?

Reply:
I can imagine a case where I want to get all possibilities instead of just the ones that follow my heuristic. In that case, I'd want to be able to toggle it off. As for the depth thing, I can't actually think of a real situation in puzzles where we wouldn't want to specify depth. I think the only reason was that [other online](http://ceptimus.co.uk/wordladder.php) [solvers](http://www.keithschwarz.com/interesting/code/word-ladder/WordLadders.html) don't take an exact depth/length as an option